### PR TITLE
Add a processor for the archival queue

### PR DESCRIPTION
**Pull request by MichaelSnowden**
**What changed?**
Added a processor which executes tasks on the archival queue. A followup PR will actually enable the queue factory.

**Why?**
I made these changes to enable the archival queue.

**How did you test it?**
I tested this using unit tests because it's not actually hooked into prod yet.

**Potential risks**
No--it's not wired into main yet.

**Is hotfix candidate?**
No.

### CODE REVIEW

File: service/history/archivalQueueTaskExecutor.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+		archiver:      archiver,
+		shardContext:  shardContext,
+		workflowCache: workflowCache,
+		logger:        logger,
+		metricsClient: metricsHandler,
+	}
+}
+
+func (e *archivalQueueTaskExecutor) Execute(ctx context.Context, executable queues.Executable) (tags []metrics.Tag,
+	isActive bool, err error) {
```
Comment (yux0) [inline service/history/archivalQueueTaskExecutor.go:63]: nit:
```
func (e *archivalQueueTaskExecutor) Execute(
    ctx context.Context,
    executable queues.Executable,
) (tags []metrics.Tag,isActive bool, err error) {
```

File: service/history/archivalQueueTaskExecutor.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+	if err != nil {
+		return err
+	}
+	defer func() { release(err) }()
+
+	mutableState, err := weContext.LoadMutableState(ctx)
+	if err != nil {
+		return err
+	}
+	if mutableState == nil || mutableState.IsWorkflowExecutionRunning() {
```
Comment (yux0) [inline service/history/archivalQueueTaskExecutor.go:92]: If the mutable state is nil, I think we should return nil to ack this task. If the workflow is deleted manually prior to archival.

File: service/history/archivalQueueTaskExecutor.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+		CallerService:        common2.HistoryServiceName,
+	})
+	if err != nil {
+		return err
+	}
+	retention := namespaceEntry.Retention()
+	if retention == 0 {
+		retention = 7 * 24 * time.Hour
+	}
+	deleteTime := closeTime.Add(retention)
```
Comment (yux0) [inline service/history/archivalQueueTaskExecutor.go:153]: Add a jitter for this

File: service/history/archivalQueueTaskExecutor.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+		VisibilityURI:        namespaceEntry.VisibilityArchivalState().URI,
+		Targets:              []archival.Target{archival.TargetHistory, archival.TargetVisibility},
+		CallerService:        common2.HistoryServiceName,
+	})
+	if err != nil {
+		return err
+	}
+	retention := namespaceEntry.Retention()
+	if retention == 0 {
+		retention = 7 * 24 * time.Hour
```
Comment (yux0) [inline service/history/archivalQueueTaskExecutor.go:151]: make it a const?

File: service/history/archivalQueueTaskExecutor.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+		WorkflowID:           task.WorkflowID,
+		RunID:                task.RunID,
+		BranchToken:          branchToken,
+		NextEventID:          nextEventID,
+		CloseFailoverVersion: closeFailoverVersion,
+		HistoryURI:           namespaceEntry.HistoryArchivalState().URI,
+		WorkflowTypeName:     workflowTypeName,
+		StartTime:            *startTime,
+		ExecutionTime:        *executionTime,
+		CloseTime:            *closeTime,
```
Comment (yux0) [inline service/history/archivalQueueTaskExecutor.go:137]: Can the archival takes in an optional of those fields so you don't need those sanity check? And if those sanity check need to be there, I think it belongs to archiver.

File: service/history/configs/config.go [hunk 6]
```diff
@@ -506,6 +508,7 @@ func NewConfig(dc *dynamicconfig.Collection, numberOfShards int32, isAdvancedVis
 
 		// Archival related
 		ArchivalTaskBatchSize:                       dc.GetIntProperty(dynamicconfig.ArchivalTaskBatchSize, 100),
+		ArchivalTaskMaxRetryCount:                   dc.GetIntProperty(dynamicconfig.ArchivalTaskMaxRetryCount, math.MaxInt32),
```
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply service/history/configs/config.go:511]: I use it in a downstream PR, but I'll remove it from this one.

File: service/history/archivalQueueTaskExecutor.go [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+	"go.temporal.io/server/service/history/tasks"
+	"go.temporal.io/server/service/history/workflow"
+)
+
+type archivalQueueTaskExecutor struct {
+	archiver      archival.Archiver
+	shardContext  shard.Context
+	workflowCache workflow.Cache
+	logger        log.Logger
+	metricsClient metrics.MetricsHandler
```
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply service/history/archivalQueueTaskExecutor.go:48]: Fixed

File: service/history/archivalQueueTaskExecutor.go [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+func newArchivalQueueTaskExecutor(archiver archival.Archiver, shardContext shard.Context, workflowCache workflow.Cache,
```
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply service/history/archivalQueueTaskExecutor.go:51]: Fixed

File: service/history/archivalQueueTaskExecutor.go (hunk 1 above)
Comment (MichaelSnowden) [reply service/history/archivalQueueTaskExecutor.go:63]: Fixed

File: service/history/archivalQueueTaskExecutor.go (hunk 2 above)
Comment (MichaelSnowden) [reply service/history/archivalQueueTaskExecutor.go:92]: Done

File: service/history/archivalQueueTaskExecutor.go (hunk 4 above)
Comment (MichaelSnowden) [reply service/history/archivalQueueTaskExecutor.go:151]: Done

File: service/history/archivalQueueTaskExecutor.go (hunk 3 above)
Comment (MichaelSnowden) [reply service/history/archivalQueueTaskExecutor.go:153]: Done in https://github.com/temporalio/temporal/pull/3565

File: service/history/archivalQueueTaskExecutor.go (hunk 5 above)
Comment (MichaelSnowden) [reply service/history/archivalQueueTaskExecutor.go:137]: Makes sense, fixed

**Comment by MichaelSnowden** [comment]
Moved to https://github.com/temporalio/temporal/pull/3663

**Pull request closed without merging**